ansible/library/action_sync.py

```diff
@@ -0,0 +1,260 @@
+    avoid deploy-mg parameter later or keep deploy-mg == True in DB.
+    - add-topo can be issues on same topo again.
+'''
+
+EXAMPLES = '''
+ - topo_name: test action sync
+    action_sync: action={{ action }} topo={{ topo_name }} dut = {{ dut }}
+        server = {{ server }} test_name={{ test_name}}
+    connection: local
+    register: action_sync
+'''
+
+MaxLogLen = 25
+
+class ActionSync(object):
+    def __init__(self):
+        self.module = AnsibleModule(
+            argument_spec=dict(
+              action=dict(required=True, type='str'),
+              topo =dict(required=True, type='str'),
+              dut=dict(required=True, type='str'),
+              server=dict(required=True, type='str'),
+              test_name=dict(required=False, type='str'),
+            ),
+            supports_check_mode=True)
+
+        self.action = self.module.params['action']
+        self.topo = self.module.params['topo']
+        self.dut = self.module.params['dut']
+        self.server = self.module.params['server']
+        self.test_name = self.module.params.get('test_name')
+
+        # class variables
+        self.file = "actionDb"
+
+        # DB & Results
+        self.db = self.readDbFile()
+
+        return
+
+    def run(self):
+        """
+            Main method of the class
+
+        """
+        try :
+            # init DB, If Not Created.
+            if len(self.db) == 0:
+                self.initDB()
+
+            if self.action == "add-topo":
+                self.processAddTopo()
+            elif self.action == "deploy-mg":
+                self.processDeploMg()
+            elif self.action == "run-tests":
+                self.processRunTests()
+            elif self.action == "remove-topo":
+                self.processRemoveTopo()
+            else:
+                raise Exception("Wrong Action:{}".format(self.action))
+
+            self.facts   =  {'action_sync': dict(self.db)}
+            self.db.close()
+            self.module.exit_json(ansible_facts=self.facts)
+        except Exception as e:
+            self.facts   =  {'action_sync': dict(self.db)}
+            self.db.close()
+            self.module.fail_json(msg="{}".format(e), ansible_facts=self.facts)
+
+        return
+
+    def processAddTopo(self):
+
+        d = self.db['current_topo']
+
+        '''
+            if None add this topo, or same topo is loaded already.
+            Later is needed if add-topo failed at later stages.
+        '''
+        if d is None or (d['topo_name'] == self.topo and d['dut'] == self.dut and \
+            d['server'] == self.server):
+
+            d = dict()
+            d['topo_name'] = self.topo
+            d['deploy-mg'] = False
+            d['dut'] = self.dut
+            d['server'] = self.server
+            self.db['current_topo'] = d
+            self.logSuccess()
+        else:
+            self.logFailure()
+            raise Exception("{} failed".format(self.action))
```

Comment:
Assume we have multiple testbeds defined in `testbed.csv`: tb01, tb02. 
If `add-topo` is executed for tb01, then run `add-topo` for tb02 would fail, right? Because db['current_topo'] is not none and db['current_topo']['topo_name'] would be `tb01`. The else branch will be hit.
If my understanding is correct, this approach won't work for a lab with multiple different test setups.